### PR TITLE
feat: follow indirect repository URLs

### LINKED ISSUE #626: Apache Indirect SCM URLs
Some Apache projects have a URL in their SCM metadata that is currently ignored by Macaron due to not appearing to be a GitHub or similar address. However, past investigation has revealed that these links are actually indirect references to GitHub addresses, as when entered into a browser a redirect is triggered. Macaron could allow such URLs by performing a GET request on the initial URL and then returning the redirect address for further analysis.

An example indirect URL:
`https://gitbox.apache.org/repos/asf?p=commons-parent.git` -> `https://github.com/apache/commons-parent`

### PR DESCRIPTION
**Pull request by benmss**
This PR allows the Repo Finder to return GitHub, etc. URLs that are initially presented as a redirect URL. URLs that can be considered valid redirects are included within a list in the defaults configuration file. URLs found from redirects are validated as per any other URL. A flag is used to prevent infinite redirect loop shenanigans.

Closes #626

### CODE REVIEW

File: src/macaron/slsa_analyzer/git_url.py [hunk 1]
```diff
@@ -616,6 +619,20 @@ def parse_remote_url(
...
+            redirect_list = defaults.get_list("repofinder", "redirect_urls", fallback=[])
+            if redirect_list and parsed_url.netloc in redirect_list:
+                response = send_get_http_raw(parsed_url.geturl(), allow_redirects=False)
+                if not response:
+                    return None
+                redirect_url = response.headers.get("location")
+                if not redirect_url:
+                    return None
+                # Pass the redirect_url through this function and prevent loops by setting the redirect parameter to an
+                # empty string.
```
Comment (nathanwn) [inline src/macaron/slsa_analyzer/git_url.py:633]: Sorry, I'm a bit confused about which empty string you are referring to here. Isn't the redirect parameter of type `bool`?

**Comment by nathanwn** [comment]
I'm unsure if redirection should be handled in the `parse_remote_url` function in the `git_url` module.

It seems like this redirection handling should only happen once at the beginning of the analysis of a target. It may be during/right after repo finding instead (which I guess may be part of your intention, confirmed by the fact that you added some redirection setting fields into the Repo Finder section in the `.ini` config).

Otherwise, I don't see why we should modify the `parse_remote_url` function. As you have mentioned previously, this function is used in different places to split a URL into different parts. Its name does not suggest that it should also handle redirection, and if that is the case then this function is doing too many things at once.

**Comment by behnazh-w** [comment]
@benmss I think this PR can be considered as a new feature.

File: src/macaron/slsa_analyzer/git_url.py (hunk 1 above)
Comment (benmss) [reply src/macaron/slsa_analyzer/git_url.py:633]: I will update this comment to match reality.

File: src/macaron/repo_finder/repo_validator.py [hunk 2]
```diff
@@ -30,3 +43,14 @@ def find_valid_repository_url(urls: Iterable[str]) -> str:
 
     # Report the first valid URL from the end of the list.
     return vcs_list.pop()
+
+
+def _resolve_redirects(parsed_url: urllib.parse.ParseResult) -> str | None:
```
Comment (tromai) [inline src/macaron/repo_finder/repo_validator.py:48]: Because we are trying to resolve the final redirected URL first before passing it to `get_remote_vcs_url`, I wonder why don't we let `requests` handles the redirection for us (as mentioned [here](https://requests.readthedocs.io/en/latest/user/quickstart/?highlight=allow_redirects#redirection-and-history)) and only care about the final URL that it lands on. In the current implementation, we only stops at the first redirection, I'm curious if there would be any case for multiple redirection in our dataset of repositories :thinking: .

Comment (tromai) [reply src/macaron/repo_finder/repo_validator.py:48]: Right, so we only supports URL with at most 1 redirection. Then I think we can leave the current implementation as it is. 
I agree that following the redirection could potentially add more cost to the operation.

File: src/macaron/slsa_analyzer/git_url.py [hunk 3]
```diff
@@ -566,6 +566,33 @@ def get_remote_vcs_url(url: str, clean_up: bool = True) -> str:
     return url_as_str
 
 
+def clean_url(url: str) -> urllib.parse.ParseResult | None:
```
Comment (tromai) [inline src/macaron/slsa_analyzer/git_url.py:569]: Because we have refactored this logic into its own function. I think it's good if we add some unit tests for it (if it's not out of scope for this PR).

File: src/macaron/repo_finder/repo_validator.py (hunk 2 above)
Comment (benmss) [reply src/macaron/repo_finder/repo_validator.py:48]: Another consideration is the security aspect, that I'll add here for the sake of completeness. Currently Macaron will only try to resolve URLs within the allowed domain list during that part of the analysis. Allowing `requests` to follow these redirects indiscriminately creates a possibility that some undesirable URL is retrieved. That said, as nothing is done with the retrieved data the threat there may still be minimal.

File: src/macaron/slsa_analyzer/git_url.py (hunk 3 above)
Comment (nathanwn) [reply src/macaron/slsa_analyzer/git_url.py:569]: Hmm, to me at least, this may not be necessary though. This function is extracted from the `parse_remote_url` function, which by itself does not have any unit tests and it is so far OK because this is a utility function called by others, each of which has unit tests already. So if we didn't have unit tests for `parse_remote_url` before and were OK with it, I wouldn't think we need unit tests for this one.

Comment (tromai) [reply src/macaron/slsa_analyzer/git_url.py:569]: I think @benmss has already added some unit tests :sweat_smile: . 
I agree with @nathanwn that unit tests are not totally necessary as this logic has been tested through unit tests of other functions that call it. However, I don't think we should revert what @benmss has added as it doesn't cause extra issues (for me). Please let me know what you think @nathanwn.   
With that being said, I will approve the changes.

**Review by tromai — APPROVED** [review]

File: src/macaron/slsa_analyzer/git_url.py (hunk 3 above)
Comment (benmss) [reply src/macaron/slsa_analyzer/git_url.py:569]: Indeed, I added some before seeing the follow up message. Probably there is no harm in keeping them now that they exist :)

File: docs/source/pages/using.rst [hunk 4]
```diff
@@ -263,10 +263,11 @@ Within the configuration file under the ``repofinder.java`` header, three option
 - ``repo_pom_paths`` (Values: List of POM tags) - Determines where to search for repository information in the POM files. E.g. scm.url.
 - ``find_parents`` (Values: True or False) - When enabled, the Repository Finding feature will also search for repository URLs in parents POM files of the current dependency.
 
-Under the related header ``repofinder``, two more options exist: ``find_repos``, and ``use_open_source_insights``:
+Under the related header ``repofinder``, three more options exist: ``find_repos``, ``use_open_source_insights``, and ``redirect_urls``:
 
 - ``find_repos`` (Values: True or False) - Enables or disables the Repository Finding feature.
 - ``use_open_source_insights`` (Values: True or False) - Enables or disables use of Google's Open Source Insights API.
+- ``redirect_urls`` (Values: List of URLs) - Found URLs that exist in this list will be returned as the URL they redirect to.
```
Comment (behnazh-w) [inline docs/source/pages/using.rst:270]: I found this sentence a bit hard to follow. How about:

```suggestion
- ``redirect_urls`` (Values: List of URLs) - These are the URLs that redirect to the actual repository URL.
```

File: src/macaron/config/defaults.ini [hunk 5]
```diff
@@ -51,6 +51,10 @@ recursive = False
 [repofinder]
 find_repos = True
 use_open_source_insights = True
+# The list of URLs that will be returned as the location they redirect to.
```
Comment (behnazh-w) [inline src/macaron/config/defaults.ini:54]: Same comment here.

File: src/macaron/slsa_analyzer/git_url.py (hunk 3 above)
Comment (behnazh-w) [reply src/macaron/slsa_analyzer/git_url.py:569]: Let's keep the test.

**Review by behnazh-w — APPROVED** [review]

File: tests/slsa_analyzer/test_git_url.py [hunk 6]
```diff
@@ -287,3 +287,21 @@ def test_parse_git_branch_output(content: str, expected_output: list[str]) -> No
...
+        ("askjdlkajsdlkajsdlkjasldjlk:scm", ""),
+    ],
+)
+def test_clean_url(url: str, expected: str) -> None:
+    """Test the functionality of the clean_url function."""
+    result = git_url.clean_url(url)
+    if result is None:
+        assert expected == ""
+    else:
+        assert expected == result.geturl()
```
Comment (nathanwn) [inline tests/slsa_analyzer/test_git_url.py:307]: Hmm, why not have the expected results as `None` instead of empty strings?

**Review by nathanwn — APPROVED** [review]
Hi. I just had one minor comment for the test. Otherwise, it looks good.

Thanks for the PR.

File: tests/slsa_analyzer/test_git_url.py [hunk 7]
```diff
@@ -287,3 +287,18 @@ def test_parse_git_branch_output(content: str, expected_output: list[str]) -> No
...
+        ("", None),
+        ("scm:git:git@github.com:FasterXML/jackson-databind.git", "git@github.com:FasterXML/jackson-databind.git"),
+        ("https://github.com/commons-io/commons-io", "https://github.com/commons-io/commons-io"),
+        ("askjdlkajsdlkajsdlkjasldjlk:scm", None),
+    ],
+)
+def test_clean_url(url: str, expected: str | None) -> None:
+    """Test the functionality of the clean_url function."""
+    result = git_url.clean_url(url)
+    assert result.geturl() == expected if result is not None else expected is None
```
Comment (behnazh-w) [inline tests/slsa_analyzer/test_git_url.py:304]: When `url` is an empty string or invalid, we should check that the `clean_url` function returns `None`. The current usage of `expected` value is strange for these cases and we could replace `""` or `None` with any arbitrary value because it doesn't really check the output of the `clean_url` function.

It would be better to split this test into two tests:

1. testing that return value of `clean_url` is `None` when expected.
2. testing that the `result.geturl()` matches the expectation when `clean_url` is not `None`.

Comment (benmss) [reply tests/slsa_analyzer/test_git_url.py:304]:
```suggestion
    assert result.geturl() == expected if expected is not None else result is None
```
Reordering the variables in the in-line if statement would seem to solve the problem, but perhaps it is better to split it still.

Comment (behnazh-w) [reply tests/slsa_analyzer/test_git_url.py:304]: I was also thinking if we could avoid splitting, but I think have two separate tests makes the properties that are tested more explicit.

**Review by behnazh-w — APPROVED** [review]
